### ISSUE TITLE
md-tooltip design flaw

### ISSUE THREAD

**Issue by djfd** [angular/material#8029]
Hi,

currently md-tooltip installs two watchers for self. why it is bad: eg. showing some table, with the only tooltip per row will add N x 2 watchers where N is rows count.

Could we rework it to be event driven only?

Thanks

**Comment by topherfangio** [comment]
https://github.com/angular/material/pull/7822/files Was just merged which should help with this. Does that satisfy your needs?

**Comment by djfd** [comment]
thanks, checking

**Comment by djfd** [comment]
hmm, instead of adding watcher in configureWatchers () you do the same in mutation observer
```
if (!scope.visibleWatcher)
+                  scope.visibleWatcher = scope.$watch('visible', onVisibleChanged );
```

so what is the difference?

IMHO, nobody changes scope.visible variable but md-tooltip code internals only, so why just do not call show/hideTooltip directly?

as for second watcher, for a direction, did you actually see an application changing it during the tooltip show time? Do we need it at all? direction change will be applied upon next tooltip show, it seems to be a reasonable compromise for the cost of being as lightweigth as possible

**Comment by clshortfuse** [comment]
@djfd I wrote the PR to lighten the watcher load. I suggest you take a second look at the code or pull request conversation since there's more technical information there, but I'll summarize.

Originally, there were two watchers per tooltip always. Now, there is a check to see if the browser supports MutationObserver. If it does, the watcher on md-direction is removed since it will always be a static value. This means there's only a watcher for md-visible. Also, this watcher is only created if the md-visible attribute is being used. If you create an md-tooltip element without said attribute, there will be no watcher.

> IMHO, nobody changes scope.visible variable but md-tooltip code internals only, so why just do not call show/hideTooltip directly?

This would definitely be something that can be used in a production environment. Of the top of my head, suppose I have a toolbar with many icons with no text. I could add a button with a question mark icon, that when clicked, forces all tooltips to show.

Regardless, again, if you don't use md-visible and have a browser that supports MutationObserver, there is no watcher usage.

**Comment by djfd** [comment]
sounds good ))

well, too rare case to have that visible watcher )) why do not use for that emit, for exmple ? but it is ok, just thinking...

please take a moment to test this as well: https://codepen.io/djFD/pen/PNRLWq try md-tooltip set to left under firefox (maybe under other browsers too, did not check

**Comment by ThomasBurleson** [comment]
Closing "tooltip flaw issue"

**Comment by clshortfuse** [comment]
I'm looking at it the codepen now.

Of course, removing the watcher would be easier, but I designed the change to be non-breaking. The title of the issue is rather vague, but the conversation seems to be mostly be related to the usage of a watcher. If you feel there's nothing wrong with md-tooltip in terms of watchers I suggest you close this issue.

You might want to open up a new issue related specifically to the tooltip positioning issue and then open up a PR to fix just that. 

I also saw https://github.com/angular/material/issues/8033. I think you based your code on v1.07 instead of master. In the future, try looking at the master branch code before starting your fixes. It'll save a bit of a headaches. Lastly, just to clarify, I'm not part of the team, but any improvements are always welcomed by us contributers as well.
